clash-prelude/src/Clash/Explicit/RAM.hs

```diff
@@ -132,32 +135,55 @@ asyncRam#
   -- ^ Value to write (at address @w@)
   -> Signal rdom a
   -- ^ Value of the @RAM@ at address @r@
-asyncRam# wclk rclk en sz rd we wr din =
-    unsafeSynchronizer wclk rclk dout
+asyncRam# !_ !_ en sz rd we wr din = dout
   where
-    rd'  = unsafeSynchronizer rclk wclk rd
     ramI = Seq.replicate
-              (snatToNum sz)
+              szI
               (withFrozenCallStack (errorX "asyncRam#: initial value undefined"))
-    en' = fromEnable (andEnable en we)
-    dout = go ramI rd' en' wr din
+    en0 = fromEnable (andEnable en we)
+    dout = if rPeriod == wPeriod
+           then goSingle ramI rd en0 wr din
+           else go 0 ramI rd en0 wr din
+    rPeriod = snatToNum (clockPeriod @rdom) :: Int
+    wPeriod = snatToNum (clockPeriod @wdom) :: Int
     szI = snatToNum sz :: Int
 
-    go :: Seq.Seq a -> Signal wdom Int -> Signal wdom Bool
-       -> Signal wdom Int -> Signal wdom a -> Signal wdom a
-    go !ram (r :- rs) (e :- es) (w :- ws) (d :- ds) =
-      let ram' = upd ram e (fromEnum w) d
+    goSingle :: Seq.Seq a -> Signal rdom Int -> Signal wdom Bool
+       -> Signal wdom Int -> Signal wdom a -> Signal rdom a
+    goSingle !ram (r :- rs) ~(e :- es) ~(w :- ws) ~(d :- ds) =
+      let ram0 = upd ram e w d
           o    = ram `safeAt` r
-      in  o :- go ram' rs es ws ds
-
-    upd ram we' waddr d = case maybeIsX we' of
-      Nothing -> case maybeIsX waddr of
-        Nothing -> fmap (const (seq waddr d)) ram
-        Just wa -> safeUpdate wa d ram
-      Just True -> case maybeIsX waddr of
-        Nothing -> fmap (const (seq waddr d)) ram
-        Just wa -> safeUpdate wa d ram
-      _ -> ram
+      in  o :- goSingle ram0 rs es ws ds
+
+    -- Given
+    --   tR = absolute time of next active edge of read clock
+    --   tW = absolute time of next active edge of write clock
+    -- relTime is defined as relTime = tW - tR
+    --
+    -- Put differently, relative time 0 points at the next active edge of the
+    -- read clock, and relTime points at the next active edge of the write
+    -- clock.
+    go :: Int -> Seq.Seq a -> Signal rdom Int -> Signal wdom Bool
+       -> Signal wdom Int -> Signal wdom a -> Signal rdom a
+    go   relTime !ram rt@(~(r :- rs)) et@(~(e :- es)) wt@(~(w :- ws))
+         dt@(~(d :- ds))
+      | relTime < 0 = let ram0 = upd ram e w d
+                      in go (relTime + wPeriod) ram0 rt es ws ds
+      | otherwise   = let o = ram `safeAt` r
+                      in o :- go (relTime - rPeriod) ram rs et wt dt
```

Comment:
Forgot to hit "start a review" :sweat_smile: 

```suggestion
                      in o :- (o `seqX` go (relTime - rPeriod) ram rs et wt dt)
```

The consumer could do this too, but in other memory blocks we take this precaution as well so it seems like a good fit here. If this is left out on purpose, could you add a comment as to why? :)

Reply:
I did the leakage assessment together with @leonschoorl , perhaps he can tell you why. I cannot. If you don't do this, do unused read outputs stay in the stream as thunks when you're further into the future?

I only looked at `blockRam#` after evaluating leakage of this function, and I noticed something else. I'm going to put that in a separate comment to take advantage of GitHub's suggestion view.

Reply:
I think we should use `defaultSeqX` here and for the data written to the memory, so people can do `-fsuper-strict` on it.

Reply:
Ah we can't do `defaultSeqX` because that requires the data to be `NFDataX` and it is not. So `seqX` it is.

Reply:
I've added the `seqX`. @leonschoorl , do you agree?

Reply:
@martijnbastiaan My reasoning for not having the ``o `seqX` `` went something like this:

Without that extra ``o `seqX` ``, if you feed that resulting signal into an register that isn't enabled (some of the time), then after taking of the head of this signal, the `o` can be garbage collected without having to be evaluated.

Now `asyncRam#` does something seemingly similar:
https://github.com/clash-lang/clash-compiler/blob/a076fa9e6fdcd0d07f547a1177e5bff28981809d/clash-prelude/src/Clash/Signal/Internal.hs#L1181-L1185
But that's subtly different.
There it forces `o` to reprevent to prevent the build up of a huge chain of thunks in the first argument of `go` that otherwise can happen when the register is held in reset.

---
Thinking about it some more and looking at `blockRam#`, forcing the `o` might be a good thing, if you see it as forcing the forcing the indexing into the (potentially large) contents for the ram.

`blockRam#` does `` o `seqX` (o :- ... )``, which make more sense to me than `` o :-  (o `seqX` ... ) ``, or is that no good because of the multiple clock domains here?

`blockRam#` also make sure to force the data being written into it, which also seems like a good idea, unless it's incompatible the the multiple domains here?

Does `asyncRam` suffer from https://github.com/clash-lang/clash-compiler/issues/1458 ? Which may need [`seqErrorX`](https://github.com/clash-lang/clash-compiler/commit/c3fe11fe2af4181c27e5de785b2c929c7fc2e856) instead of `seqX`.

As for the `seqX` vs `defaultSeqX`, you could just add the `NFDataX` constraint, (almost?) any other storage element has that anyway: `register`, `blockRam`, `delay`, ....

Reply:
> Thinking about it some more and looking at `blockRam#`, forcing the `o` might be a good thing, if you see it as forcing the forcing the indexing into the (potentially large) contents for the ram.

So the indexing operation itself seems pretty low computational overhead if the output is not used, but the same might not hold for the read address, which necessarily also will be evaluated. It's not really clear to me why you would want to force `o` to WHNF here, or to quote myself: _If you don't do this, do unused read outputs stay in the stream as thunks when you're further into the future?_ I don't know.

> `blockRam#` does `` o `seqX` (o :- ... )``, which make more sense to me than `` o :-  (o `seqX` ... ) ``, or is that no good because of the multiple clock domains here?

For the single clock domain version, that seems fine, but for the multiple clock domain version I'm not so sure indeed. Producing the constructor first seems like the safer option.

> `blockRam#` also make sure to force the data being written into it, which also seems like a good idea, unless it's incompatible the the multiple domains here?

I [do that](https://github.com/clash-lang/clash-compiler/pull/2006/commits/4c7f11326268bbdfec5561ecf5183bf63afed0f0#diff-42184b38e50f74675cd2af9c3fec2656e3036f76967a5c98a846df0bb37dbf46R188) as well.

> Does `asyncRam` suffer from https://github.com/clash-lang/clash-compiler/issues/1458 ? Which may need [`seqErrorX`](https://github.com/clash-lang/clash-compiler/commit/c3fe11fe2af4181c27e5de785b2c929c7fc2e856) instead of `seqX`.

No, out-of-bounds indexing evaluates the _whole `Signal`_ to `XException` rather than a sample in the `Signal`. I think this is wrong (as we discussed on Slack), but for now, I'd say it is handled by `seqX`. However, I'll add a unit test that tests this case, so CI will punish us if we ever make that mistake (for instance when we change OOB to produce `ErrorCall`).

> As for the `seqX` vs `defaultSeqX`, you could just add the `NFDataX` constraint, (almost?) any other storage element has that anyway: `register`, `blockRam`, `delay`, ....

Ah, this comment made me realise I had forgotten the `backport 1.4` label. Since I'd like to backport this fix to 1.4, I don't think we should add constraints in this PR. Perhaps in a separate PR that is not backported.

Reply:
> No, out-of-bounds indexing evaluates the _whole `Signal`_ to `XException` rather than a sample in the `Signal`. 

I'm only half right. This is what happens for OOB write address and asserted write enable. And as discussed on Slack, it is defensible to stop the simulation in such a case. For OOB read address, it's currently the sample that is an `XException`, not the `Signal`.

Anyway, I added a unit test that tests for #1458.

Reply:
>> `blockRam#` does `` o `seqX` (o :- ... )``, which make more sense to me than `` o :-  (o `seqX` ... ) ``, or is that no good because of the multiple clock domains here?

>For the single clock domain version, that seems fine [edit] Hmmm not so sure anymore, man this is intricate [/edit]

Yes, it's fine, I just tested it with a loop that would get stuck otherwise.